### PR TITLE
[clang] Optimize castToDeclContext for 2% improvement in build times

### PR DESCRIPTION
**Pull request by Destroyerrrocket**
[copied from issue]

The function castToDeclContext takes around 2% of the execution time in my test run.

I was profiling clang to see if by any chance I could spot any small mistakes that were taking a significant amount of time on debug builds, and while pretty much everything is either complicated enough where optimizing would be impossible for me or is already very performant, I noticed that valgrind was reporting an interesting function as the top of the 'auto' category:
interesting_entry

It is being executed 1.8 billion times, and the implementation looks pretty trivial to me:
code_snippet

Looking at the assembly, I expected pretty much a lookup table and an add operation, but it is pretty clear that the resulting code is not ideal:

```asm
clang::Decl::castFromDeclContext(clang::DeclContext const*): # @clang::Decl::castFromDeclContext(clang::DeclContext const*)
.L_ZN5clang4Decl19castFromDeclContextEPKNS_11DeclContextE$local:
  movzwl 8(%rdi), %edx
  movq %rdi, %rax
  andl $127, %edx
  leal -1(%rdx), %esi
  cmpl $84, %esi
  ja .LBB65_7
  leaq .LJTI65_0(%rip), %rdi
  movq $-40, %rcx
  movslq (%rdi,%rsi,4), %rsi
  addq %rdi, %rsi
  jmpq *%rsi
.LBB65_2:
  addq %rcx, %rax
  retq
.LBB65_4:
  movq $-48, %rcx
  addq %rcx, %rax
  retq
.LBB65_5:
  movq $-64, %rcx
  addq %rcx, %rax
  retq
.LBB65_6:
  movq $-56, %rcx
  addq %rcx, %rax
  retq
.LBB65_7:
  leal -53(%rdx), %esi
  movq $-72, %rcx
  cmpl $6, %esi
  jb .LBB65_2
  addl $-34, %edx
  xorl %ecx, %ecx
  cmpl $5, %edx
  setae %cl
  shll $4, %ecx
  orq $-64, %rcx
  addq %rcx, %rax
  retq
.LJTI65_0:
<JUMP TABLE>
```

The PR I'll submit in a few minutes fixes this problem by eliminating the need for the macro DECL_CONTEXT_BASE (it's only used here and in two other analogous functions), and reordering the AST decl order to prioritize classes that inherit from DeclContext. I also experimented with hand rolled offset tables, but this is far from maintainable even if it manages to compress 3 lookup tables into one. The resulting assembly is just:

```asm
clang::Decl::castFromDeclContext(clang::DeclContext const*): # @clang::Decl::castFromDeclContext(clang::DeclContext const*)
.L_ZN5clang4Decl19castFromDeclContextEPKNS_11DeclContextE$local:
  movzwl 8(%rdi), %ecx
  leaq .Lswitch.table._ZN5clang4Decl19castFromDeclContextEPKNS_11DeclContextE(%rip), %rdx
  movq %rdi, %rax
  andl $127, %ecx
  addq (%rdx,%rcx,8), %rax
  retq
```
  
And the build difference of clang+clang-tools-extra with a debug build:
NonOpt: ninja  19007,02s user 760,01s system 2284% cpu 14:25,23 total
Opt: ninja  18806,18s user 763,33s system 2308% cpu 14:07,74 total

So around ~1.02 speedup, ~0.98 of the previous execution, nothing earth shattering, but what would be expected from valgrind, and I already did all the legwork, so I might as well send it :)

### CODE REVIEW

**Comment by nikic** [comment]
https://llvm-compile-time-tracker.com/compare.php?from=98624914367bf9091919de330cf322fb6d5e468f&to=7d398e29b0180aacc646fed5406ad7fbd4b30c02&stat=instructions%3Au Nice results for C++ code!

**Comment by Destroyerrrocket** [comment]
Yeah, I also saw that in O3 most of the compile time is overshadowed by LLVM, so it makes perfect sense that the -O0 are the ones showing better results!

**Review by shafik — COMMENTED** [review]
Thank you for your PR, this does look like a nice performance improvement, I added some reviewers who might have more detailed comments.

Please add a detailed summary for this PR. The summary is what usually goes into the git log and it is important that the git log is useful for quick understanding of changes without having to dig into the details.

**Comment by Destroyerrrocket** [comment]
Copied description from the issue that tracks this PR (sorry, I'm away from my PC)

**Review by Endilll — COMMENTED** [review]
Core idea of removing `DECL_CONTEXT_BASE` and range checks for `DeclContext` nodes seems correct to me, as all this machinery wasn't intended for `DeclContext` in the first place. This is apparent from TableGen definition, which lacks any base parameter:
https://github.com/llvm/llvm-project/blob/463dad107f4cb60ae1d49138143d6797599fb1fb/clang/include/clang/Basic/DeclNodes.td#L10

But changes made to `ClangASTNodesEmitter.cpp` are not as clear to me. I'd expect it to simply opt out `DeclContext` nodes from some code generation, but instead I see a lot of changes concerned with `PrioritizedClasses`. It would be nice if you explain what's going on in that file.

**Comment by Destroyerrrocket** [comment]
The reason for puting the classes with DeclContext closer together is to achieve a better code generation. Here's an example of the asssembly with just the removal of the DECL_CONTEXT_BASE macro:
```
clang::Decl::castToDeclContext(clang::Decl const*): # @clang::Decl::castToDeclContext(clang::Decl const*)
.L_ZN5clang4Decl17castToDeclContextEPKS0_$local:
  movl 28(%rdi), %edx
  leaq .LJTI66_0(%rip), %rsi
  movq %rdi, %rax
  movl $40, %ecx
  andl $127, %edx
  decl %edx
  movslq (%rsi,%rdx,4), %rdx
  addq %rsi, %rdx
  jmpq *%rdx
.LBB66_5:
  movl $48, %ecx
  addq %rcx, %rax
  retq
.LBB66_1:
  addq %rcx, %rax
  retq
.LBB66_2:
  movl $72, %ecx
  addq %rcx, %rax
  retq
.LBB66_6:
  movl $64, %ecx
  addq %rcx, %rax
  retq
.LBB66_7:
  movl $56, %ecx
  addq %rcx, %rax
  retq
.LBB66_8:
.LJTI66_0:
<jump table>
```
While this is better, it is still not performing ideally compared to just the load and an add of the final implementation. I'm of course open to suggestions if there is a better way of getting llvm to emit the right thing.

I agree with @cor3ntin that a comment explaining the need for this order priorization is needed.

File: clang/utils/TableGen/ClangASTNodesEmitter.cpp
Comment (Destroyerrrocket) [reply clang/utils/TableGen/ClangASTNodesEmitter.cpp:106]: I'll make this a lambda

File: clang/utils/TableGen/ClangASTNodesEmitter.cpp
Comment (Destroyerrrocket) [reply clang/utils/TableGen/ClangASTNodesEmitter.cpp:231]: Conformity with surrounding code. I code like I see to not introduce another style. I can switch to std::string_view if you wish so!

File: clang/utils/TableGen/ClangASTNodesEmitter.cpp
Comment (Destroyerrrocket) [reply clang/utils/TableGen/ClangASTNodesEmitter.cpp:267]: I copied over the Tree implementation a few lines up with some code removed. I will update both for loops, it certainly would improve readability.

**Comment by Destroyerrrocket** [comment]
Addressed the first review comments! Thank you for your time, I appreciate it :)

File: clang/utils/TableGen/ClangASTNodesEmitter.cpp
Comment (Endilll) [inline clang/utils/TableGen/ClangASTNodesEmitter.cpp:77]:
```suggestion
    auto vecPrioritized = PriorizeIfSubclassOf.empty()
```

**Review by Endilll — APPROVED** [review]
LGTM, but you should wait for more approvals before merging.

**Review by cor3ntin — APPROVED** [review]
LGTM.
1% is a very consequential improvement, congrats!

**Comment by Destroyerrrocket** [comment]
Thank you! I truly appreciate the congrats! :D
@Endilll I don't have write access (I think? I'm new here :) ), I think one of you will have to make the merge (not used to github, at work I use bitbucket, I might be wrong, I'm basing this on "Only those with [write access](https://docs.github.com/articles/what-are-the-different-access-permissions) to this repository can merge pull requests.")

**Comment by Endilll** [comment]
@Destroyerrrocket We'll merge as soon as you deem it ready and provide commit description. PR description doesn't seem a good fit for a commit description to me.

**Comment by Destroyerrrocket** [comment]
Squashed, rebased, and updated the commit message to give more information. You can go ahead and merge!

**Pull request merged**
